lib/core/ring/cne_ring.c

```diff
@@ -172,38 +176,61 @@ cne_ring_create(const char *name, unsigned int esize, unsigned int count, unsign
     if (flags & RING_F_EXACT_SZ)
         count = cne_align32pow2(count + 1);
 
-    ring_size = cne_ring_get_memsize_elem(((esize == 0) ? RING_DFLT_ELEM_SZ : esize), count);
+    ring_size = cne_ring_get_memsize_elem(esize, count);
     if (ring_size < 0) {
-        errno = EINVAL;
+        errno = -ring_size;
         return NULL;
     }
 
-    /* Reserve a memory region for this ring
-       Calloc can return pointer which is not cache aligned.
-       In that case cne_ring_init calling memset will segfault on clang compiled
-       with -O3. Tested with clang-10.0.0-4ubuntu1 and gcc-9.3.0-10ubuntu2
-    */
-    ring_mem = calloc(1, ring_size + CNE_CACHE_LINE_SIZE - 1);
+    if (addr) {
+        /* Address must be aligned to a cacheline boundary */
+        if (!cne_is_aligned(addr, CNE_CACHE_LINE_SIZE)) {
+            errno = EINVAL;
+            CNE_NULL_RET("ring is not cache aligned r=%p aligned=%p\n", addr,
+                         CNE_PTR_ALIGN(addr, CNE_CACHE_LINE_SIZE));
+        }
+
+        /* Use the supplied buffer as the ring buffer and structure memory */
+        if (ring_size > size) {
+            errno = ENOMEM;
+            return NULL;
+        }
+        ring_mem = addr;
+        memset(ring_mem, 0, ring_size);
+    } else {
+        /* Reserve a memory region for this ring
+         * Calloc can return pointer which is not cache aligned.
+         * In that case cne_ring_setup calling memset will segfault on clang compiled
+         * with -O3. Tested with clang-10.0.0-4ubuntu1 and gcc-9.3.0-10ubuntu2
+         */
+        ring_mem = calloc(1, ring_size + CNE_CACHE_LINE_SIZE - 1);
```

Comment:
Instead of checking for ring_mem != NULL later, why not check it here and save yourself from needing an extra if statement later. Also, it seems odd to set flags |= RING_F_ALLOCATED if the memory allocation fails (i know free can take a NULL address, but still, it seems off).

Reply:
added the suggested change.